Description: fix FTBFS against libpng1.6
Author: Tobias Frost <tobi@debian.org>
Bug-Debian: https://bugs.debian.org/742560
Last-Update: 2016-01-25

--- a/Source/FreeImage/PluginPNG.cpp
+++ b/Source/FreeImage/PluginPNG.cpp
@@ -713,11 +713,19 @@

                       if (png_get_valid(png_ptr, info_ptr, PNG_INFO_iCCP)) {
                               png_charp profile_name = NULL;
+#if PNG_LIBPNG_VER_MAJOR >= 1 && PNG_LIBPNG_VER_MINOR >= 4
+                              png_bytepp profile_data = NULL;
+#else
                               png_charp profile_data = NULL;
+#endif
                               png_uint_32 profile_length = 0;
                               int  compression_type;

+#if PNG_LIBPNG_VER_MAJOR >= 1 && PNG_LIBPNG_VER_MINOR >= 4
+                              png_get_iCCP(png_ptr, info_ptr, &profile_name, &compression_type, profile_data, &profile_length);
+#else
                               png_get_iCCP(png_ptr, info_ptr, &profile_name, &compression_type, &profile_data, &profile_length);
+#endif

                               // copy ICC profile data (must be done after FreeImage_AllocateHeader)

@@ -1000,7 +1008,11 @@

                       FIICCPROFILE *iccProfile = FreeImage_GetICCProfile(dib);
                       if (iccProfile->size && iccProfile->data) {
+#if PNG_LIBPNG_VER_MAJOR >= 1 && PNG_LIBPNG_VER_MINOR >= 4
+                              //png_set_iCCP(png_ptr, info_ptr, "Embedded Profile", 0, (png_const_bytep)iccProfile->data, iccProfile->size);
+#else
                               png_set_iCCP(png_ptr, info_ptr, "Embedded Profile", 0, (png_charp)iccProfile->data, iccProfile->size);
+#endif
                       }

                       // write metadata
